modules/saslserv/main.c

```diff
@@ -739,6 +705,64 @@ sasl_mech_unregister(struct sasl_mechanism *mech)
 	}
 }
 
+static bool
+sasl_authcid_can_login(struct sasl_session *const restrict p, const char *const restrict authcid,
+                       myuser_t **const restrict muo)
+{
+	myuser_t *const mu = myuser_find_by_nick(authcid);
+
+	if (! mu)
+		return false;
+
+	if (muo)
+		*muo = mu;
+
+	p->authceid = sstrdup(entity(mu)->id);
+
+	hook_user_login_check_t req = {
+
+		.si         = p->si,
+		.mu         = mu,
+		.allowed    = true,
+	};
+
+	(void) hook_call_user_can_login(&req);
+
+	if (! req.allowed)
+		(void) logcommand(p->si, CMDLOG_LOGIN, "failed LOGIN to \2%s\2 (denied by hook)", entity(mu)->name);
+
+	return req.allowed;
+}
+
+static bool
+sasl_authzid_can_login(struct sasl_session *const restrict p, const char *const restrict authzid,
```

Comment:
I'll admit the code duplication between `sasl_authzid_can_login` and `sasl_authcid_can_login` makes me a little unhappy.